### DISCUSSION TITLE
C# wrapper

### DISCUSSION THREAD

**Discussion by tonyvfish** [free-audio/clap#277] [Q&A]
I see that the header file(s) have been ported to Delphi, just wondering if the same thing is possible for .Net / C#.  I can of course translate the header .into c#, but is that enough, or would it need a unmanaged wrapper C that call a managed one ?
TIA

**Comment by domportera** [comment]
I would love if a proper C#/.NET wrapper was available.

If the GC really is a deal breaker (imo it probably shouldnt be since you can still minimize GC), [Bb exists](https://flattened.net/), and the "Zero" version doesn't even have a GC

**Reply by tonyvfish** [reply]
I wasn't aware of BB, that looks really interesting. As you say the GC can be suppressed or has low latency modes. I built a small DAW using https://github.com/obiwanjacobi/vst.net and did not experience GC issues ?

I've been looking at Avalonia where you can easily build cross platform UIs too. Just wondering if you need the managed stub to make the Clap plugin?

**Comment by abique** [comment]
Did you consider writing the engine in C/C++/Rust and then run the GUI in a child process which you can write in C#?

**Reply by tonyvfish** [reply]
No I didn't , but thank you for the hint Alexandre. I'm a very experienced c# developer but my C skills are very rusty. If you think I completely managed application is not feasible I will look into that. Is a completely c# managed application possible ?

**Reply by baconpaul** [reply]
So @tonyvfish there is nothing stopping you writing a fully managed app, but it may not have the performance characteristics your users expect, which is why many folks use a mix of managed and unmanaged. 

In the simplest form, realtime audio apps like DAWs have "two" threads, a "main" thread which is the UI thread and an "audio" thread which runs all the audio and mixing and stuff. You can see in the clap headers there are annotations for which thread a function is called on.

The reason people are hesitant to use GC languages is because of that audio thread. To rough approximation, that audio thread has to generate 64 samples every 64/48000th of a second and if it is late, you get a pop or click. So running a GC on the audio thread (or, really, even running any allocation or deletion of memory) is something people don't do. 

But for a very large number of the activities (save state, manage windows, assemble messages to deliver to the audio thread, draw your UI) which happen on the main thread, the GC languages are perfectly fine.

So that leads to an architecture where you write your 'main' thread in a GC language and it does things like scan and create and initialize objects and so on; and you write your 'audio' thread in a non-GC environment and it calls process; and you use lock free data structures to send messages between your main and audio threads. Several DAWs work this way (bitwig, carla I'm pretty sure, a couple of others) because writing a complex UI in c++ is miserable but running the gc on your audio thread is bad.

Pragmatically if you write your entire app including the audio thread in C# it will "probably" work for you "most" of the time as it will for "most" of your users, but that occasional pop and drop is a problem.

If you are not writing a realtime app, though (like if you are writing a composing tool which pre-renders audio), none of this matters. Buffer ahead. But for anything with performance and realtime gestures you do have to think about it quite a lot.

Hope that helps!

**Reply by abique** [reply]
As a side note, Bitwig Studio is written in Java, and its audio engine in C++.
They both run in different processes.

**Reply by baconpaul** [reply]
yeah the managed program is sort of a window viewer on the unmanaged audio engine. makes debugging a dream too!

**Reply by baconpaul** [reply]
(like seriously; you can restart your plugin engine without resetting your entire ui state)

**Reply by tonyvfish** [reply]
I suppose what I was getting at without looking to deeply at the API is, when the vsthost calls a plugin it's must expect at least one entry point in the plugin dll to be invoked? Does that entry point have to be defined in an unmanaged language?

**Reply by tonyvfish** [reply]
I've just had a look at the docs and I can see just one entry point, export symbol is required and I also came across this https://sites.google.com/site/robertgiesecke/Home/uploads/unmanagedexports which you create C dll style exports from . Net.  Thank you all for your suggestions and advice about managed garbage collected languages. Actually the GC is very sophisticated now, you can suppress it entirely or put it into a low latency mode, so I may give that a go, or as someone else advised use
Bb https://flattened.net/ the unmanaged version of c#, either way you have provided me lots of options, thank you

**Comment by domportera** [comment]
[BEEF](https://www.beeflang.org/) could be a really excellent alternative to C# for this kind of thing - quality of life improvements of C# with no JIT/GC and out-of-the-box simple interop with C and C++.

**Reply by abique** [reply]
There seem to be a new language every month these days.

**Reply by abique** [reply]
Interesting.

**Reply by domportera** [reply]
for real. this one has been around for some time now and seems pretty ready for real work. it's a shame it's not supported by any IDEs but it has its own at least. why someone would bother making their own IDE for this is beyond me, but it seems like a worthwhile project

**Comment by tonyvfish** [comment]
I'm currently working on a basic c# version of a plugin using .Net7 . You can do some great things now regarding interop and GC https://github.com/AaronRobinsonMSFT/DNNE

**Reply by domportera** [reply]
how is it that you're using this repo? what's your general architecture?

**Reply by tonyvfish** [reply]
> how is it that you're using this repo? what's your general architecture?

Sorry for the late reply, I have currently been using this project as a reference for the types i need to create in C#
The idea is to convert this example plugin by Moss
 [Moss Example Plugin](https://github.com/git-moss/clap-tutorials) into  C#
The aim is to pass all the CLAP validation that Moss's plugin does

To help me do the basic Marshalling Interop i've used this project
[interop asisstant](https://github.com/jaredpar/pinvoke-interop-assistant)

And to create the entry points in the C# i've used this
[ProtoTypeExports](https://github.com/AaronRobinsonMSFT/DNNE)

The ideas is the main C# .Net project is invoked from an unmanaged stub dll plugin.  This is because DNNE ProtoTypeExports, support export functions and not a single entry point variable structure.
This first C# project does the marshalling and works with pointers etc, almost a pseudo C project.  This project will be the C# CLAP plugin framework.  A consumer will make reference to this project and will create a plugin discoverable by the C# CLAP plugin framework.
The framework will then delegate its calls in a safe way to the users plugin

In this way the Plugin implementer will we working with interfaces and will not have to concern themselves with pointer or marshalling
I am currently at the stage of passing the CLAP validator test
I need to then do some dog fooding and make a real example plugin.
I will post ASAP

As [abique](https://github.com/abique) mentioned , this does not necessarily mean a managed language is the correct thing to do a plugin in.  It suitability will depend on the purpose and the user control of garbage collection and being aware of efficient operations like Span<T> ans Memory<T>

**Reply by baconpaul** [reply]
the clap-saw-demo at <https://github.com/surge-synthesizer/clap-saw-demo> is a pretty robust micro synth which uses the clap api directly and may be a useful next port if you are looking for examples; It supports polymod, note expressions, and so on.

**Reply by tonyvfish** [reply]
> the clap-saw-demo at https://github.com/surge-synthesizer/clap-saw-demo is a pretty robust micro synth which uses the clap api directly and may be a useful next port if you are looking for examples; It supports polymod, note expressions, and so on.

wow thank you Paul

**Comment by atsushieno** [comment]
You might want to reach people behind NPlug project (https://github.com/xoofx/NPlug , https://elk.zone/g0v.social/@xoofx@mastodon.social/109978124900894558 )

I had very similar arguments to this thread there.
